### PR TITLE
ascan: Make addition of query param optional

### PR DESCRIPTION
**Pull request by kingthorin**
Following feedback via stackoverflow, google groups, the issue tracker etc. make the addition of a query parameter to GET requests optional.

See also: https://github.com/zaproxy/zaproxy/pull/2863

I'll prepare a related core help PR after this is accepted. (https://github.com/zaproxy/zap-core-help/wiki/HelpUiDialogsOptionsAscaninput)

### CODE REVIEW

File: src/org/parosproxy/paros/core/scanner/ScannerParam.java [hunk 1]
```diff
@@ -159,6 +168,17 @@
...
+    /**
+     * Flag that indicates if a query param should be added to GET requests that don't have one.
+     * <p>
+     * Default value is {@code false}.
+     * 
+     * @since TODO add version
+     * @see #isAddQueryParam()
+     * @see #setAddQueryParam(boolean)
+     */
+    private boolean addQueryParam;
```
Comment (thc202) [inline src/org/parosproxy/paros/core/scanner/ScannerParam.java]: It's missing the read in the `parse()` method.

File: src/org/zaproxy/zap/extension/ascan/OptionsVariantPanel.java [hunk 2]
```diff
@@ -104,21 +105,24 @@ private JPanel getPanelScanner() {
             panelInjectable.add(
                     this.getChkInjectableQueryString(), 
                     LayoutHelper.getGBC(0, 1, 1, 1.0D, 0, GridBagConstraints.HORIZONTAL, new Insets(2, 8, 2, 2)));            
+            panelInjectable.add(
+                    this.getChkAddQueryParam(),
+                    LayoutHelper.getGBC(0, 2, 1, 1.0D, 0, GridBagConstraints.HORIZONTAL, new Insets(2, 32, 2, 2)));
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascan/OptionsVariantPanel.java]: Please, extract a variable for the row, less changes next time :)

File: src/org/zaproxy/zap/extension/ascan/OptionsVariantPanel.java [hunk 3]
```diff
@@ -335,10 +344,29 @@ private JCheckBox getChkInjectableQueryString() {
         if (chkInjectableQueryString == null) {
             chkInjectableQueryString = new JCheckBox();
             chkInjectableQueryString.setText(Constant.messages.getString("variant.options.injectable.querystring.label"));
+            chkInjectableQueryString.addItemListener(new ItemListener() {
+
+                @Override
+                public void itemStateChanged(ItemEvent e) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascan/OptionsVariantPanel.java]: Could use a lambda, less code.

**Comment by thc202** [comment]
The test `VariantAbstractQueryUnitTest` needs to be updated.

**Comment by kingthorin** [comment]
Review addressed.

**Comment by thc202** [comment]
Thank you!

**Comment by kingthorin** [comment]
I guess this should have an equivalent set of API calls. Shall I update this PR or do it separately?

**Comment by thc202** [comment]
They are added automatically, but worth adding the descriptions.

**Comment by kingthorin** [comment]
Done

File: src/org/parosproxy/paros/core/scanner/ScannerParam.java [hunk 4]
```diff
@@ -213,6 +233,8 @@ protected void parse() {
 
         this.scanHeadersAllRequests = getBoolean(SCAN_HEADERS_ALL_REQUESTS, false);
 
+        this.addQueryParam = getBoolean(SCAN_ADD_QUERY_PARAM, false);
```
Comment (psiinon) [inline src/org/parosproxy/paros/core/scanner/ScannerParam.java]: Wouldnt it be better to have this as on by default to maintain the current behaviour?

Comment (kingthorin) [reply src/org/parosproxy/paros/core/scanner/ScannerParam.java]: I don't mind changing it, I went false due to the feedback/attention it's been getting.

Comment (kingthorin) [reply src/org/parosproxy/paros/core/scanner/ScannerParam.java]: Anyone else got feedback on this? Shall I start a User Group thread?

Comment (thc202) [reply src/org/parosproxy/paros/core/scanner/ScannerParam.java]: I'm OK changing the default given the reported FP rate. We could mention the change in the release notes, to be more noticeable.

Comment (kingthorin) [reply src/org/parosproxy/paros/core/scanner/ScannerParam.java]: Agreed, does that seems reasonable to you @psiinon?

Comment (kingthorin) [reply src/org/parosproxy/paros/core/scanner/ScannerParam.java]: Help PR updated: https://github.com/zaproxy/zap-core-help/pull/203

Comment (psiinon) [reply src/org/parosproxy/paros/core/scanner/ScannerParam.java]: Yeah, but we should def mention this in the release notes.

**Comment by kingthorin** [comment]
Ref: https://mobile.twitter.com/kingthorin_rm/status/1041855009263108097

**Pull request merged**